napari_clusters_plotter/_clustering.py

```diff
@@ -323,56 +326,112 @@ def reset_choices(self, event=None):
-            # write result back to features/properties of the labels layer
-            add_column_to_layer_tabular_data(
-                labels_layer, "HDBSCAN_CLUSTER_ID_SCALER_" + str(standardize), y_pred
-            )
-        else:
-            warnings.warn(
-                "Clustering unsuccessful. Please check again selected options."
-            )
-            return
+        run_clustering(
+            labels_layer=labels_layer,
+            selected_measurements_list=selected_measurements_list,
+            selected_method=selected_method,
+            **kwargs,
+        )
 
         # show region properties table as a new widget
         from ._utilities import show_table
 
         show_table(self.viewer, labels_layer)
 
 
+def run_clustering(
+    labels_layer: Labels,
+    selected_measurements_list: list,
+    selected_method: str,
+    num_clusters: int = 2,
+    num_iterations: int = 100,
+    standardize: bool = True,
+    min_cluster_size: int = 5,
+    min_nr_samples: int = 5,
+):
+    """
+    Run selected clustering algorithm on data stored in a layer.properties dict.
+
+    Parameters
+    ----------
+    labels_layer : Labels
+        labels_layer with attribute `properties` or `features`
+    selected_measurements_list : list
+        list of features that should be included for the clustering. Must be
+        list of strings (e.g., `['measurement_1', 'measurement_2', ...]`).
+    selected_method : str
+        Selected clustering method. Can be either of ['KMEANS', 'HDBSCAN'].
+    num_clusters : int, optional
+        Number of clusters to be detected. Only applicable to PCA clustering.
+        The default is 2.
+    num_iterations : int, optional
+        Number of iterations for clustering procedure Only applicable to PCA.
+        The default is 100.
+    standardize : bool, optional
+        Whether to standardize the data with z-score normalization. The default
+        is True.
+    min_cluster_size : int, optional
+        Minimal amount of data points to constitute a cluster. Only applicable
+        to HDBSCAN. The default is 5.
+    min_nr_samples : int, optional
+        Measure of how conservative the algorithm should decide on whether to
+        add a data point to a cluster or not. The default is 5.
+
+    Returns
+    -------
+    None.
+
+    See also
+    --------
+    https://hdbscan.readthedocs.io/en/latest/index.html
+    https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
+    """
+    print("Selected labels layer: " + str(labels_layer))
+    print("Selected measurements: " + str(selected_measurements_list))
+    print("Selected clustering method: " + str(selected_method))
+
+    features = get_layer_tabular_data(labels_layer)
+
+    # only select the columns the user requested and remove NaNs
+    selected_properties = features[selected_measurements_list]
+    non_nan_entries = features.index  # take all entries here
```

Comment:
Hey @jo-mueller, since Robert changed the code for clustering already I would merge that PR first and then implement this change you are suggesting on that branch, since I think it makes more sense to keep Roberts changes, as he is adding functionality there!

Reply:
Hey @Cryaaa , agree, makes a lot of sense. Let's revisit this PR later when merging the other one has been finished, I think we will run into some merge conflicts anyway.